[image-picker][Android] Rewrite module to Sweet API

bbarthec opened the pull request:
# Why

Partially resolves [ENG-3071](https://linear.app/expo/issue/ENG-3071/audit-expo-image-picker-android-rewrite-to-sweet-api)
Extracted part of #16251

# How

1. Convert module to `expo-module`
    - add `android` section to `expo-modules.config.json`
    - remove `ImagePickerPackage` and replace it with `ImagePickerModule`
2. `ImagePickerModule` is implements `Module`
    - add `definition` with all JS exported methods
3. create `ImagePickerOptions` and `ImagePickerResponse` implementing `Record`
4. unify `Exceptions` and make them implement `CodedException`
5. migrate to `kotlin.promise`
    - add proxy method to legacy `Permissions` API
6. clean up:
    - refactored `EXIF_TAGS` (from `Map<"string" | "double" | "int", ExifValue>` to `Map { "string" to List<ExifValue", "double" to List<ExifValue>, "int" to List<ExifValue> }`)
    - converting `quality` value from 0.0-1.0 (coming from JS API) to 0-100 (acceptable by Android bitmap exporters) i pushed down to the exporters

# Test Plan

- I was using `ImagePickerScreen` from `ncl` in `bare-expo` app

# To be delivered in separate PR:
- [ ] multiselection
- [ ] using `registerForActivityResult` awaiting #17572 to be merged
- [ ] bumping `com.github.CanHub:Android-Image-Cropper` to the newest version
- [ ] fir the problem when OS kills the `Activity` due to resources limit and the results are lost (even though there's a mechanism handling that, but it seems not to work properly)

bbarthec committed:
[image-picker][Android] Make `expo-image-picker` visible as `expo-module`

bbarthec committed:
[image-picker][Android] Convert `ImagePickerOptions` to `Record`

bbarthec committed:
[image-picker][Android] Create `ImagePickerResponse`

Create `ImagePickerResponse` that implements `Record` to allow auto-conversion using Sweet API

bbarthec committed:
[image-picker][Android] Unify all exceptions

bbarthec committed:
[image-picker][Android] Refactor constants

- remove all constant strings that are now encapsulated either in Options/Response or Exceptions
- change `EXIF_TAGS` structure from `Map<"string" | "double" | "int", ExifValue>` to `Map { "string" to List<ExifValue", "double" to List<ExifValue>, "int" to List<ExifValue> }`

bbarthec committed:
[image-picker][Android] Migrate to `Options/Response`, `Exceptions` and `kotlinPromise`

- refactor `Image/VideoResultTask` to use new `Options`/`Response` and `Exceptions`

bbarthec committed:
[image-picker][Android] Remove old `ImagePickerPackage`

bbarthec committed:
[image-picker][Android] Refactor `ImagePickerModule` to adapt SweetAPI

- migrate to `kotlin.Promise`
- create module definition
- create convinient `context` property
- adapt `Options`/`Response` and `Exceptions`

bbarthec committed:
[e-m-c][Android] Expose methods accepting `kotlin.Promise`

bbarthec committed:
[skip ci] Update CHANGELOG

bbarthec committed:
[image-picker][Android] `./gradlew spotlessApply`

bbarthec committed:
[image-picker][Android] Fix imports in `Expo Go`

lukmccall commented on packages/expo-image-picker/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 
 ### 💡 Others
 
+- Native module on Android is now written in Swift using [Sweet API](https://docs.expo.dev/modules/module-api). ([#17668](https://github.com/expo/expo/pull/17668) by [@bbarthec](https://github.com/bbarthec))
```
```suggestion
- Native module on Android is now written in Kotlin using [Sweet API](https://docs.expo.dev/modules/module-api). ([#17668](https://github.com/expo/expo/pull/17668) by [@bbarthec](https://github.com/bbarthec))
```
However, it would be fun to see a Android code written in Swift 😜

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerResponse.kt:
```diff
@@ -0,0 +1,43 @@
+  @Field
+  val cancelled: Boolean = false
+
+  class Image(
```
I like how did you use a nested class with a sealed class on the top 🔥

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt:
```diff
@@ -12,9 +12,13 @@ import java.io.IOException
 class CompressionImageExporter(
   private val mImageLoader: ImageLoaderInterface,
-  private val mQuality: Int,
+  private val mQuality: Double,
```
```suggestion
 mQuality: Double,
```
It doesn't have to be stored inside this class. Also, I think we should rename it to not start with `m`.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt:
```diff
@@ -12,9 +12,13 @@ import java.io.IOException
 
 class CompressionImageExporter(
   private val mImageLoader: ImageLoaderInterface,
```
```suggestion
  private val imageLoader: ImageLoaderInterface,
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt:
```diff
@@ -12,9 +12,13 @@ import java.io.IOException
   private val mImageLoader: ImageLoaderInterface,
-  private val mQuality: Int,
+  private val mQuality: Double,
   private val mBase64: Boolean
```
```suggestion
  private val base64: Boolean
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/tasks/ImageResultTask.kt:
```diff
@@ -108,16 +98,17 @@ open class ImageResultTask(
       val exifInterface = ExifInterface(input)
-      exifTags.forEach { (type, name) ->
-        if (exifInterface.getAttribute(name) != null) {
+      EXIF_TAGS.flatMap { (type, tags) -> tags.map { tag -> type to tag } }
```
Hmm, this line is very magical 🧙‍♀️

From the context, it is very easy to know what it's doing, but I would put it inside the helper function to make it more readable.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
 
+class ImagePickerModule : Module() {
+  private val context: Context
+    get() = appContext.reactContext.ifNull { throw IllegalStateException("React Application Context is null. You can't access the Context at this point.") }
```
I know that we have `ifNull`, but in that case, we can also use `?:`, but  I don't have a strong opinion here.
```suggestion
    get() = appContext.reactContext ?: throw IllegalStateException("React Application Context is null. You can't access the Context at this point.")
```

However, I would change the `IllegalStateException` exception to the custom one (`CodedException`). In the future, we may have logic which will handle it.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
+   * Flag indicating that the main activity (host) was killed while performing cropping.
    */
-  private var mWasHostDestroyed = false
+  private var mWasHostDestroyedWhileCropping = false
```
```suggestion
  private var wasHostDestroyedWhileCropping = false
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
 
-  //region expo methods
+    AsyncFunction("launchCameraAsync") { options: ImagePickerOptions, promise: Promise ->
+      val activity = appContext.currentActivity.ifNull {
+        throw MissingCurrentActivityException()
+      }
```
```suggestion
      val activity = appContext.currentActivity ?: throw MissingCurrentActivityException()
```
up to you

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
-  fun requestMediaLibraryPermissionsAsync(writeOnly: Boolean, promise: Promise) {
-    Permissions.askForPermissionsWithPermissionsManager(mPermissions, promise, *getMediaLibraryPermissions(writeOnly))
-  }
+      val permissions = appContext.permissions.ifNull {
+        throw ModuleNotFoundException("permissions")
+      }
```
```suggestion
      val permissions = appContext.permissions ?: throw ModuleNotFoundException("permissions")
```
up to you

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
-  fun getMediaLibraryPermissionsAsync(writeOnly: Boolean, promise: Promise) {
-    Permissions.getPermissionsWithPermissionsManager(mPermissions, promise, *getMediaLibraryPermissions(writeOnly))
-  }
+      val intentType = if (options.mediaTypes == MediaTypes.VIDEOS) MediaStore.ACTION_VIDEO_CAPTURE else MediaStore.ACTION_IMAGE_CAPTURE
```
```suggestion
      val intentType = if (options.mediaTypes == MediaTypes.VIDEOS) {
        MediaStore.ACTION_VIDEO_CAPTURE 
      } else { 
        MediaStore.ACTION_IMAGE_CAPTURE
      }
```

lukmccall reviewed (commented):
🔥🔥🔥

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt:
```diff
@@ -32,174 +21,177 @@ import expo.modules.imagepicker.fileproviders.CacheFileProvider
-    }
-  }
+  private fun getMediaLibraryPermissions(writeOnly: Boolean): Array<String> =
+    if (writeOnly) arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    else arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
```
Afaik we skip braces only in one-liners
```suggestion
    if (writeOnly) { 
      arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    } else { 
      arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
    }
```

bbarthec committed:
Apply code review suggestions

Cleanup all `m<XXX>` prefixes
Get rid of `ifNull` in favour of `:?` (elvis operator)
Add bracers to multiline `if else` statements
Get rid of storing some values as propoerties in `<X>ImageExporter`

bbarthec committed:
Refactor `EXIF_TAGS` to conform `Iterable` interface

bbarthec committed:
[emc][skip CI] Update CHANGELOG

bbarthec committed:
[skip ci] ./gradlew :expo-image-picker:spotlessApply
